micro-rdk/src/common/data_manager.rs

```diff
@@ -262,43 +252,128 @@ where
+                // MAX_SENSOR_CONTENTS_SIZE
+                let upload_success = loop {
+                    let next_message = match reader.read_next_message() {
+                        Ok(msg) => msg,
+                        Err(err) => {
+                            log::error!(
+                                "error reading message from store for collector key ({:?}): {:?}",
+                                collector_key,
+                                err
+                            );
+                            // we don't want to panic, and creating an AppClientError variant for this case
+                            // feels too specific, so we'll return to exit the function early
+                            break false;
+                        }
+                    };
+                    if next_message.len() > MAX_SENSOR_CONTENTS_SIZE {
+                        log::error!(
+                            "message encountered that was too large (>32K bytes) for collector {:?}",
+                            collector_key
+                        );
+                        continue;
+                    }
+                    let current_chunk_size: usize = current_chunk.iter().map(|m| m.len()).sum();
+                    if next_message.is_empty()
+                        || ((next_message.len() + current_chunk_size) > MAX_SENSOR_CONTENTS_SIZE)
+                    {
+                        let data: Result<Vec<SensorData>, prost::DecodeError> =
+                            current_chunk.drain(..).map(SensorData::decode).collect();
+                        let data = match data {
+                            Ok(data) => data,
+                            Err(err) => {
+                                log::error!(
+                                    "error decoding readings for collector key ({:?}): {:?}",
+                                    collector_key,
+                                    err
+                                );
+                                break false;
+                            }
+                        };
+                        let upload_request = DataCaptureUploadRequest {
+                            metadata: Some(UploadMetadata {
+                                part_id: self.part_id.clone(),
+                                component_type: collector_key.component_type.clone(),
+                                r#type: DataType::TabularSensor.into(),
+                                component_name: collector_key.r_name.clone(),
+                                method_name: collector_key.method.to_string(),
+                                ..Default::default()
+                            }),
+                            sensor_contents: data,
+                        };
+                        match app_client
+                            .upload_data(upload_request)
+                            .or(async {
```

Comment:
Just confirming my understanding: we will have the lock on the store during this upload, and that is desired.

Reply:
yes, the timeout will ensure that we are not holding on to the lock for very long

Reply:
Let's add a comment making clear that we are intentionally doing the RPC with the lock held and why we believe it is safe to do so.